### PR TITLE
scanner: should not add the same wwn device again

### PR DESCRIPTION
**Pull request by Vicente-Cheng**
**Problem:**
Get multiple blockdevice with the same WWN

**Solution:**
To prevent data broken with mis-operation, we should prevent creating the blockdevice with the same WWN.

**Related Issue:**
https://github.com/harvester/harvester/issues/4494

**Test plan:**
1. Create scsi controller files
- controller.xml
```
    <controller type='scsi' model='virtio-scsi' index='0'/>
```
Then, use following command to add virtio-scsi controller
```
virsh attach-device --domain <VM name> --file <controller file name> --live
```

2. Prepare two disk XML files with the same wwn as below:
- node-0-sda.xml
```
    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2'/>
      <source file='/tmp/node-0-sda.qcow2'/>
      <target dev='sda' bus='scsi'/>
      <wwn>0x5000c50017654321</wwn>
    </disk>
```

- node-0-sdb.xml
```
    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2'/>
      <source file='/tmp/node-0-sdb.qcow2'/>
      <target dev='sdb' bus='scsi'/>
      <wwn>0x5000c50017654321</wwn>
    </disk>
```

3. Then try to add these two disks to the host.
You can use the following command to add disk manually.
```
virsh attach-device --domain <VM name> --file <disk xml file> --live
```

4. Check the blockdevices CR. It only has one blockdevice existing, and it should be the first added.

**TODO ITEM**
- [x] Improve controller
- [x] add integration

### CODE REVIEW

**Comment by Vicente-Cheng** [comment]
Hi folks, this PR is also ready to review.
I added the CI integration test for duplicated wwn disks.

Please help to check it. Thanks! @bk201, @tserong

File: tests/integration/test_1_disk_hotplug_test.go [hunk 1]
```diff
@@ -148,6 +150,43 @@ func (s *HotPlugTestSuite) Test_2_HotPlugAddDisk() {
 	require.Equal(s.T(), err, nil, "Get Blockdevices should not get error")
 
 	require.Equal(s.T(), curBlockdevice.Status.State, diskv1.BlockDeviceActive, "Disk status should be inactive after we add disk")
+	s.curBusPath = curBlockdevice.Status.DeviceStatus.Details.BusPath
+}
+
+func (s *HotPlugTestSuite) Test_3_AddDeplicatedWWNDsik() {
```
Comment (tserong) [inline tests/integration/test_1_disk_hotplug_test.go:156]: Small typo fix:

```suggestion
func (s *HotPlugTestSuite) Test_3_AddDuplicatedWWNDisk() {
```

**Review by tserong — APPROVED** [review]
LGTM, I just found one small typo.

Question: what is it that creates `/tmp/hotplug_disks/node1-sda.xml` in the first place in the automated tests? I can't figure that out.

File: tests/integration/test_1_disk_hotplug_test.go (hunk 1 above)
Comment (Vicente-Cheng) [reply tests/integration/test_1_disk_hotplug_test.go]: nice catch! update!

**Comment by Vicente-Cheng** [comment]
>Question: what is it that creates /tmp/hotplug_disks/node1-sda.xml in the first place in the automated tests? I can't figure that out.

` /tmp/hotplug_disks/node1-sda.xml` and ` /tmp/hotplug_disks/node1-sda.qcow2` are created by script attach-disk.sh on the previous stage. You can refer to this https://github.com/harvester/node-disk-manager/blob/master/.github/workflows/basic-ci.yaml#L54-L59

**Comment by tserong** [comment]
> ` /tmp/hotplug_disks/node1-sda.xml` and ` /tmp/hotplug_disks/node1-sda.qcow2` are created by script attach-disk.sh on the previous stage. You can refer to this https://github.com/harvester/node-disk-manager/blob/master/.github/workflows/basic-ci.yaml#L54-L59

Got it, thanks @Vicente-Cheng!

File: pkg/controller/blockdevice/scanner.go [hunk 2]
```diff
@@ -131,7 +132,7 @@ func (s *Scanner) scanBlockDevicesOnNode() error {
 		return err
 	}
 
-	oldBds := convertBlockDeviceListToMap(oldBdList)
+	oldBds, createdWWNs := convertBlockDeviceListToMap(oldBdList)
```
Comment (bk201) [inline pkg/controller/blockdevice/scanner.go:135]:
```suggestion
	oldBds, existingWWNs := convertBlockDeviceListToMap(oldBdList)
```

File: pkg/controller/blockdevice/scanner.go [hunk 3]
```diff
@@ -179,13 +189,15 @@ func (s *Scanner) scanBlockDevicesOnNode() error {
 	return nil
 }
 
-func convertBlockDeviceListToMap(bdList *diskv1.BlockDeviceList) map[string]*diskv1.BlockDevice {
+func convertBlockDeviceListToMap(bdList *diskv1.BlockDeviceList) (map[string]*diskv1.BlockDevice, []string) {
+	createdWWNs := make([]string, 0)
```
Comment (bk201) [inline pkg/controller/blockdevice/scanner.go:193]:
```suggestion
	wwns := make([]string, len(bdlist))
```

Comment (Vicente-Cheng) [reply pkg/controller/blockdevice/scanner.go]: OK, let's keep the capability to a fixed length.
Thanks for the suggestion.

File: pkg/controller/blockdevice/scanner.go (hunk 2 above)
Comment (Vicente-Cheng) [reply pkg/controller/blockdevice/scanner.go]: updated!

**Review by bk201 — APPROVED** [review]
LGTM, big thanks!

**Comment by Vicente-Cheng** [comment]
@mergifyio backport v0.5.x

**Pull request merged**